### PR TITLE
Fix 'attempting to reference a deleted function' and gtest static crt

### LINKED ISSUE #8192: MSVC Error, attempting to reference a deleted function
I am trying to rebuild chainerx with latest everything (os, compiler, cuda/cudnn),
I also modified the `setup.py` trying to build with ninja generator. And encounter the following error:
```
C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1423~1.281\bin\Hostx64\x64\cl.exe  /nologo /TP -DCHAINERX_ENABLE_CUDA=1 -DCHAINERX_ENABLE_LAPACK=1 -DWIN32_LEAN_AND_MEAN -Dgsl_api="" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -Ipybind11\include -Igsl-lite\include -Iabseil -Igoogletest-src\googletest\include -ID:\chainer\chainerx_cc -DVERSION_INFO=\"7.0.0b4\"  /DWIN32 /D_WINDOWS /W3 /GR /EHsc /openmp /MD /O2 /Ob2 /DNDEBUG   -std:c++14 /showIncludes /Fochainerx\routines\CMakeFiles\chainerx_routines.dir\connection.obj /Fdchainerx\routines\CMakeFiles\chainerx_routines.dir\chainerx_routines.pdb /FS -c D:\chainer\chainerx_cc\chainerx\routines\connection.cc
D:\chainer\chainerx_cc\chainerx\routines\connection.cc(64): warning C4244: 'argument': conversion from 'int64_t' to 'int8_t', possible loss of data
D:\chainer\chainerx_cc\chainerx\routines\connection.cc(365): error C2280: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': attempting to reference a deleted function
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: see declaration of 'chainerx::BackwardBuilder::BackwardBuilder'
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': function was explicitly deleted
```

https://github.com/chainer/chainer/blob/f653368d7f296878360f3d98ff1c128fb2a51e1d/chainerx_cc/chainerx/routines/connection.cc#L365-L366
where 
https://github.com/chainer/chainer/blob/f653368d7f296878360f3d98ff1c128fb2a51e1d/chainerx_cc/chainerx/backward_builder.h#L141

The error seems to be valid.

### PR DESCRIPTION
**Pull request by cloudhan**
Fixed #8192 

simply make it build with msvc

### CODE REVIEW

**Comment by cloudhan** [comment]
Jenkins, test this please.

File: chainerx_cc/CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,7 @@ option(CHAINERX_ENABLE_LAPACK "Use LAPACK if available" ${DEFAULT_CHAINERX_ENABL
 if(MSVC)
     option(CUDA_USE_STATIC_CUDA_RUNTIME "Use the static version of the CUDA runtime library if available" OFF)
     set(CUDA_USE_STATIC_CUDA_RUNTIME OFF CACHE INTERNAL "")
+    add_definitions(-DWIN32_LEAN_AND_MEAN)
```
Comment (niboshi) [inline chainerx_cc/CMakeLists.txt]: Why is this needed?

File: chainerx_cc/chainerx/routines/connection.cc [hunk 2]
```diff
@@ -362,8 +362,9 @@ Array Linear(const Array& x, const Array& w, const absl::optional<Array>& b, uin
         }
     }
 
-    BackwardBuilder bb = has_bias ? BackwardBuilder{"linear", {x_matrix, w, b_matrix}, out_matrix}
-                                  : BackwardBuilder{"linear_nobias", {x_matrix, w}, out_matrix};
+    BackwardBuilder bb{has_bias ? "linear" : "linear_nobias",
+                       has_bias ? std::vector<ConstArrayRef>{x_matrix, w, b_matrix} : std::vector<ConstArrayRef>{x_matrix, w},
+                       out_matrix};
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: Hmm, this does not look very elegant...
Is there any other way to fix the error?

Comment (cloudhan) [reply chainerx_cc/chainerx/routines/connection.cc]: https://github.com/chainer/chainer/blob/f653368d7f296878360f3d98ff1c128fb2a51e1d/chainerx_cc/chainerx/backward_builder.h#L141
make it `= default;` Will also fix this, granted.

File: chainerx_cc/CMakeLists.txt (hunk 1 above)
Comment (cloudhan) [reply chainerx_cc/CMakeLists.txt]: Hmm, leak from #8194, sorry for this, will remove.

File: chainerx_cc/chainerx/routines/connection.cc (hunk 2 above)
Comment (cloudhan) [reply chainerx_cc/chainerx/routines/connection.cc]: OK, I will keep this for now. I tried `BackwardBuilder& operator=(BackwardBuilder&&) = default;` But I still need to std::move here...😂

Comment (niboshi) [reply chainerx_cc/chainerx/routines/connection.cc]: I think it's still better even if `std::move` is required. In the current code a single conditional is scattered into multiple expressions, which does not look elegant.
How do you think?

Comment (cloudhan) [reply chainerx_cc/chainerx/routines/connection.cc]: Default move assignment is not needed anyway. Just move here will be OK. But default move constructor with `noexcept` makes MSVC unhappy.

Comment (niboshi) [reply chainerx_cc/chainerx/routines/connection.cc]: OK, please fix as such.
As for `noexcept`, it is required to suppress clang-tidy error.
Please feel free to report an issue though.

**Comment by cloudhan** [comment]
@niboshi travis-ci says core dump on test_linear.py but I cannot reproduce it on my linux machine, with Python 3.5/3.7 + cuda 10.0

it says
```
23791 passed, 4411 skipped, 56650 deselected, 111 xfailed in 325.84 seconds
```

Any suggestion?

**Comment by cloudhan** [comment]
Maybe the ternary move trigger the same bug in gcc mentioned in
https://stackoverflow.com/questions/53374182/ternary-operator-with-xvalue-operands
which caused core dump in https://travis-ci.org/chainer/chainer/builds/592439020.
Anyway, I cannot reproduce the segfault in my local windows/linux machine.

File: chainerx_cc/chainerx/backward_builder.h [hunk 3]
```diff
@@ -136,7 +136,7 @@ class BackwardBuilder {
     ~BackwardBuilder() { CHAINERX_ASSERT(is_finalized_); }
 
     BackwardBuilder(const BackwardBuilder&) = delete;
-    BackwardBuilder(BackwardBuilder&&) noexcept = default;
+    BackwardBuilder(BackwardBuilder&&) = default;
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward_builder.h]: Why is this needed?
`noexcept` was requested by clang-tidy: https://clang.llvm.org/extra/clang-tidy/checks/performance-noexcept-move-constructor.html

Comment (cloudhan) [reply chainerx_cc/chainerx/backward_builder.h]: needed by d15fcad, no longer for now

**Comment by cloudhan** [comment]
OK, in summary, to fix 'attempting to reference a deleted function'
1. current master branch code 
https://github.com/chainer/chainer/blob/cfaa1053c15383410b90af3f09b807d0c3771cda/chainerx_cc/chainerx/routines/connection.cc#L365-L366
MSVC is unhappy

2. code in 2b367ea
https://github.com/chainer/chainer/blob/2b367ead8de9c514ec0abd898ae01e748d18f50f/chainerx_cc/chainerx/routines/connection.cc#L365-L366
and 
https://github.com/chainer/chainer/blob/2b367ead8de9c514ec0abd898ae01e748d18f50f/chainerx_cc/chainerx/backward_builder.h#L139
MSVC is happy, while GCC happily compile the code with bug which cause segfault.

3.
https://github.com/chainer/chainer/blob/89c88336742dbedd446f9c98e7208bf5f07e253c/chainerx_cc/chainerx/routines/connection.cc#L365-L367
is a bit ugly, niboshi is unhappy, all compilers and test case are happy

**Comment by niboshi** [comment]
OK. Do you have any resource to help understand the problem in MSVC?
I want some code comment here, because otherwise someone might be tempted to "clean" the code to the previous version.

The link you posted above doesn't seem to explain the problem enough.
If you do not, maybe a link to this PR is enough.

**Comment by cloudhan** [comment]
> OK. Do you have any resource to help understand the problem in MSVC?
> I want some code comment with a link here, because otherwise someone might be tempted to "clean" the code to the previous version.
> 
> The link you posted above doesn't seem to explain the problem enough.
> If you do not, maybe a link to this PR is enough.

Maybe quote to the issue #8192?  Might be standard conformance problem for MSVC. But I have tested adding `/permissive-` flags and removing `/openmp` (currently, MSVC make these two mutually exclusive), no chance. The move assignment is deleted, so MSVC is correct, this is the case 1. Case 2 makes MSVC compiles. But in CI with gcc build, it cause a core dump. I tried to reproduce it, no luck, MSVC and GCC builds all pass the the test case. Case 3 is simply a compromise.

**Comment by niboshi** [comment]
@cloudhan 
>Maybe quote to the issue #8192? 

That seems to work as well. Could you write a comment?

File: chainerx_cc/chainerx/routines/connection.cc [hunk 4]
```diff
@@ -362,8 +362,12 @@ Array Linear(const Array& x, const Array& w, const absl::optional<Array>& b, uin
         }
     }
 
-    BackwardBuilder bb = has_bias ? BackwardBuilder{"linear", {x_matrix, w, b_matrix}, out_matrix}
-                                  : BackwardBuilder{"linear_nobias", {x_matrix, w}, out_matrix};
+    // NOTE: 1. BackwardBuilder's move assignment is deleted.
+    //       2. GCC does not work well with move ctor with ternary move or move ternary as argument.
+    //       3. please refers to issue 8192 and it related PR for more information.
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: It looks the most important piece of information, that MSVC fails to compile the original code, is missing.

Also, please write the link as `#8192`.

**Comment by niboshi** [comment]
By the way, what happens if move assignment operator is defined (`= default;`)?

**Comment by cloudhan** [comment]
> By the way, what happens if move assignment operator is defined (`= default;`)?

msvc complains it will weirdly call another deleted function with const BackwardBuilder&

**Comment by niboshi** [comment]
Can you post the error message?

**Comment by cloudhan** [comment]
```
D:\chainer\chainerx_cc\chainerx\routines\connection.cc(365): error C2280: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': attempting to reference a deleted function
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: see declaration of 'chainerx::BackwardBuilder::BackwardBuilder'
D:\chainer\chainerx_cc\chainerx/backward_builder.h(138): note: 'chainerx::BackwardBuilder::BackwardBuilder(const chainerx::BackwardBuilder &)': function was explicitly deleted
```
image

error of ternary move:
image

and error of move ternary 
image

which is also mentioned in https://stackoverflow.com/questions/53374182/ternary-operator-with-xvalue-operands (the previous link), it calls the copy ctor.

Can we stop struggling with this trivial problem anymore?

**Comment by cloudhan** [comment]
to be even clearer, with 
image
no error is generated with the intellisense, but same error is generated.

even this doesn't work, same error.
image

To make MSVC accept above cases, you need to also remove `noexcept` from the move ctor

**Comment by cloudhan** [comment]
```c++
static_assert(std::is_nothrow_constructible<BackwardBuilder, const char*, std::vector<ConstArrayRef>, const Array&>::value);
```
will fail.

File: chainerx_cc/chainerx/routines/connection.cc [hunk 5]
```diff
@@ -362,8 +362,12 @@ Array Linear(const Array& x, const Array& w, const absl::optional<Array>& b, uin
         }
     }
 
-    BackwardBuilder bb = has_bias ? BackwardBuilder{"linear", {x_matrix, w, b_matrix}, out_matrix}
-                                  : BackwardBuilder{"linear_nobias", {x_matrix, w}, out_matrix};
+    // NOTE: 1. BackwardBuilder's move assignment is deleted, use it will cause compilation error with MSVC.
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: I asked about move assignment operator because of this comment.

So, you mean defining the move assignment operator does not change error message at all, right?
(comparing with the original report https://github.com/chainer/chainer/issues/8192#issue-499526799).

Comment (cloudhan) [reply chainerx_cc/chainerx/routines/connection.cc]: yeah

**Comment by cloudhan** [comment]
Let's stopping worrying about this PR for now. I'll improve it latter. @niboshi, just leave alone for now.

File: chainerx_cc/chainerx/routines/connection.cc (hunk 5 above)
Comment (niboshi) [reply chainerx_cc/chainerx/routines/connection.cc]: >BackwardBuilder's move assignment is deleted,

Hm, then this clause sounds irrelevant.

**Comment by niboshi** [comment]
@cloudhan OK.
